core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala

```diff
@@ -401,12 +539,35 @@ class ReassignPartitionsCommand(zkUtils: ZkUtils, proposedAssignment: Map[TopicA
       allProposed.filter { case (tp, _) => tp.topic == topic })
   }
 
-  def reassignPartitions(throttle: Throttle = NoThrottle): Boolean = {
+  def reassignPartitions(throttle: Throttle = NoThrottle, timeoutMs: Long = 10000L): Boolean = {
     maybeThrottle(throttle)
     try {
-      val validPartitions = proposedAssignment.filter { case (p, _) => validatePartition(zkUtils, p.topic, p.partition) }
+      val validPartitions = proposedPartitionAssignment.filter { case (p, _) => validatePartition(zkUtils, p.topic, p.partition) }
       if (validPartitions.isEmpty) false
       else {
+        if (proposedReplicaAssignment.nonEmpty) {
+          // Send AlterReplicaDirRequest to allow broker to create replica in the right log dir later if the replica
+          // has not been created it. This allows us to rebalance load across log directories in the cluster even if
+          // we can not move replicas between log directories on the same broker. We will be able to move replicas
+          // between log directories on the same broker after KIP-113 is implemented.
+          val adminClient = adminClientOpt.getOrElse(
+            throw new AdminCommandFailedException("bootstrap-server needs to be provided in order to reassign replica to the specified log directory"))
+          val alterReplicaDirResult = adminClient.asInstanceOf[KafkaAdminClient].alterReplicaDir(
```

Comment:
Good point. I added the following comments in the code:

```
/*
 * Before KIP-113 is fully implemented, user can only specify the destination log directory of the replica
 * if the replica has not already been created on the broker; otherwise the log directory specified in the
 * json file will not be enforced. Therefore we want to verify that broker will return ReplicaNotAvailableException
 * for this replica.
 *
 * After KIP-113 is fully implemented, we will not need to verify that the broker returns this ReplicaNotAvailableException
 * in this step. And after the reassignment znode is created, we will need to repeated send AlterReplicaDirRequest to broker
 * if broker returns ReplicaNotAvailableException for any replica in the request.
 */
```

And I updated the help message for option `--reassignment-json-file` to the following:

```
The JSON file with the partition reassignment configurationThe format to use is -                         

{"partitions":                        
	[{"topic": "foo",                    
	  "partition": 1,                    
	  "replicas": [1,2,3],               
	  "log_dirs": ["dir1","dir2","dir3"] 
  }],                                 
"version":1                           
}                                     

Note that "log_dirs" is optional. When it is specified, its length must equal the length of the replicas list. The value in this list can be either "any" or the absolution path of the log directory on the broker.  If absolute log directory path is specified, it is currently required that the replica has not already been created on that broker. The replica will then be created in the specified log directory on the broker later.                       

```